File: support/src/julia/utils/ConfigUtils.jl
"""
ConfigUtils - Configuration file utilities.

Provides utilities for loading and managing TOML configuration files.

# Exports
- `load_toml_config`: Load a TOML configuration file
- `get_config_value`: Safely get a nested config value
- `merge_configs`: Merge multiple config dictionaries
"""
module ConfigUtils

using TOML

export load_toml_config, get_config_value, merge_configs

"""
    load_toml_config(path::AbstractString; required::Bool=true) -> Dict{String,Any}

Load a TOML configuration file.

# Arguments
- `path`: Path to the TOML file
- `required`: If true, throw error if file doesn't exist

# Returns
- Dictionary of configuration values
"""
function load_toml_config(path::AbstractString; required::Bool=true)::Dict{String,Any}
    if !isfile(path)
        if required
            error("Configuration file not found: $path")
        else
            return Dict{String,Any}()
        end
    end
    return TOML.parsefile(path)
end

"""
    get_config_value(config::Dict, keys...; default=nothing)

Safely get a nested configuration value.

# Example
```julia
config = load_toml_config("config.toml")
value = get_config_value(config, "section", "key", default=10)
```
"""
function get_config_value(config::Dict, keys...; default=nothing)
    current = config
    for key in keys
        if !haskey(current, key)
            return default
        end
        current = current[key]
    end
    return current
end

"""
    merge_configs(configs::Dict...) -> Dict{String,Any}

Merge multiple configuration dictionaries. Later configs override earlier ones.
"""
function merge_configs(configs::Dict...)::Dict{String,Any}
    result = Dict{String,Any}()
    for config in configs
        merge!(result, config)
    end
    return result
end

end # module

File: support/src/python/utils/config_utils.py
"""
config_utils - Configuration file utilities.

Provides utilities for loading and managing configuration files.
"""

import json
import logging
from pathlib import Path
from typing import Any, Dict, Optional

logger = logging.getLogger(__name__)

# Try to import tomllib (Python 3.11+) or toml
try:
    import tomllib
    def _load_toml(path: str) -> Dict[str, Any]:
        with open(path, 'rb') as f:
            return tomllib.load(f)
except ImportError:
    try:
        import toml
        def _load_toml(path: str) -> Dict[str, Any]:
            return toml.load(path)
    except ImportError:
        def _load_toml(path: str) -> Dict[str, Any]:
            raise ImportError("Neither tomllib (Python 3.11+) nor toml package is available")


def load_toml(path: str, required: bool = True) -> Dict[str, Any]:
    """
    Load a TOML configuration file.
    
    Args:
        path: Path to the TOML file
        required: If True, raise error if file doesn't exist
        
    Returns:
        Dictionary of configuration values
    """
    if not Path(path).exists():
        if required:
            raise FileNotFoundError(f"Configuration file not found: {path}")
        return {}
    
    return _load_toml(path)


def load_json(path: str, required: bool = True) -> Dict[str, Any]:
    """
    Load a JSON configuration file.
    
    Args:
        path: Path to the JSON file
        required: If True, raise error if file doesn't exist
        
    Returns:
        Dictionary of configuration values
    """
    if not Path(path).exists():
        if required:
            raise FileNotFoundError(f"Configuration file not found: {path}")
        return {}
    
    with open(path, 'r') as f:
        return json.load(f)


def get_config_value(config: Dict[str, Any], *keys, default: Any = None) -> Any:
    """
    Safely get a nested configuration value.
    
    Args:
        config: Configuration dictionary
        *keys: Keys to traverse
        default: Default value if key not found
        
    Returns:
        The configuration value or default
        
    Example:
        >>> config = {'section': {'key': 'value'}}
        >>> get_config_value(config, 'section', 'key', default='default')
        'value'
    """
    current = config
    for key in keys:
        if not isinstance(current, dict) or key not in current:
            return default
        current = current[key]
    return current


def merge_configs(*configs: Dict[str, Any]) -> Dict[str, Any]:
    """
    Merge multiple configuration dictionaries.
    
    Later configs override earlier ones. Performs deep merge for nested dicts.
    
    Args:
        *configs: Configuration dictionaries to merge
        
    Returns:
        Merged configuration dictionary
    """
    result = {}
    
    for config in configs:
        for key, value in config.items():
            if key in result and isinstance(result[key], dict) and isinstance(value, dict):
                result[key] = merge_configs(result[key], value)
            else:
                result[key] = value
    
    return result
